src/typechecker/FStar.TypeChecker.TcTerm.fst

```diff
@@ -462,13 +462,31 @@ let guard_letrecs env actuals expected_c : list (lbname*typ*univ_names) =
                 (Print.binders_to_string ", " bs) (Print.comp_to_string c);
 
           //exclude types and function-typed arguments from the decreases clause
+          //and reveal and erased arguments
           let filter_types_and_functions (bs:binders)  =
-            bs |> List.collect (fun ({binder_bv=b}) ->
-                    let t = N.unfold_whnf env (U.unrefine b.sort) in
-                    match t.n with
-                        | Tm_type _
-                        | Tm_arrow _ -> []
-                        | _ -> [S.bv_to_name b]) in
+            let out_rev, env = 
+              List.fold_left 
+                (fun (out, env) binder ->
+                  let b = binder.binder_bv in
+                  let t = N.unfold_whnf env (U.unrefine b.sort) in
+                  let env = Env.push_binders env [binder] in
+                  match t.n with
+                  | Tm_type _
+                  | Tm_arrow _ -> 
+                    (out, env)
+                  | _ ->
+                    let arg = S.bv_to_name b in
+                    let arg =
+                      match is_erased_head t with
+                      | Some (u, ty) -> U.apply_reveal u ty arg
```

Comment:
The check is very syntactic: If the head constructor of the argument's type is `erased`, then insert a reveal in the decreases clause